### PR TITLE
Refactor build system to use CMake

### PR DESCRIPTION
**Pull request by erimatnor**
Moving the build system to CMake allows easy cross-platform
compiles, dependency checks, and more. In particular, CMake
allows us to easily build on Windows, and Visual Studio now
has native CMake support.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,13 +98,21 @@ We recommend following our detailed [installation instructions](http://docs.time
...
 ```bash
 git clone git@github.com:timescale/timescaledb.git
+cd timescaledb
+
+# Bootstrap the build system
+./bootstrap
 
 # To build the extension
-make
+cd build && make
```
Comment (cevian) [inline README.md]: Why do we need a separate build dir? Is this standard cmake?

File: sql/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+# Command and target for the SQL file
+add_custom_command(
+  OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${EXT_SQL_FILE}
+  DEPENDS ${SQL_FILES}
+  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+  COMMAND ${CAT_SQL_FILE_CMD}
+  COMMENT "Generating ${EXT_SQL_FILE}")
+
+add_custom_target(sqlfile ALL DEPENDS ${EXT_SQL_FILE})
```
Comment (cevian) [inline sql/CMakeLists.txt]: Does add_custom target know that ${EXT_SQL_FILE} is actually  ${CMAKE_CURRENT_BINARY_DIR}/${EXT_SQL_FILE}?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,133 @@
+set(CMAKE_C_FLAGS_DEBUG "-DUSE_ASSERT_CHECKING=1")
+
+if (UNIX)
+  set(CMAKE_SHARED_LINKER_FLAGS "-L${PG_LIBDIR}")
+  set(CMAKE_MODULE_LINKER_FLAGS "-L${PG_LIBDIR}")
+  set(CMAKE_C_FLAGS "${PG_CFLAGS}")
+  set(CMAKE_CPP_FLAGS "${PG_CPPFLAGS}")
+  set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -g")
```
Comment (cevian) [inline src/CMakeLists.txt]: Are we sure it safe/correct to add debugging to our so even if the main postgres lib was compiled w/o debug symbols?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+  COMMAND ${CMAKE_COMMAND} -E echo_append "#define EXT_GIT_COMMIT " > ${GITCOMMIT_H}
+  COMMAND ${GIT_EXECUTABLE} describe --abbrev=4 --dirty --always --tags >> ${GITCOMMIT_H} || echo "nocommit" >> ${GITCOMMIT_H}
+  COMMENT "Generating gitcommit.h"
+  VERBATIM)
+
+add_library(${PROJECT_NAME} MODULE ${SOURCES} ${HEADERS} ${GITCOMMIT_H})
+
+set_target_properties (${PROJECT_NAME} PROPERTIES
+  OUTPUT_NAME ${PROJECT_NAME}-${PROJECT_VERSION_MOD}
```
Comment (cevian) [inline src/CMakeLists.txt]: I'd rather not change the so name from  `timescaledb` to `timescaledb-0.5.0` in this PR. I don't see a reason to do it here and the sql files still reference timescaledb.so. What happens under things that aren't unix? Should be left to separate PR....

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+set(UPDATE_FROM_VERSION ${CMAKE_MATCH_4})
+
+if (VERSION_MOD)
+  set(PROJECT_VERSION_MOD ${VERSION}-${VERSION_MOD})
+else ()
+  set(PROJECT_VERSION_MOD ${VERSION})
+endif ()
+
+if (NOT CMAKE_BUILD_TYPE)
+  # Default to Debug builds
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply CMakeLists.txt]: We don't really have a default now. Release or debug depends on how the Postgres build was configured as the relevant flags are configured from `pg_config`. This is true even with CMake, but this allows as to get debug symbols for the from the extension even if Postgres wasn't built with debug.

If you prefer to default to release builds, however, I am fine with changing that.

[the remark this answers is not in the document]
Comment (erimatnor) [reply CMakeLists.txt]: Ehh, actually, I guess you're referring to the mismatch between the comment and the setting, and that the actual setting is in fact release? Yeah, that should be fixed. Still, do we want to default to release or debug? I figured release was safest.

File: README.md (hunk 1 above)
Comment (erimatnor) [reply README.md]: I think out-of-source tree builds are encouraged with CMake, but not strictly necessary. Think this is a cleaner default though.

File: sql/CMakeLists.txt (hunk 2 above)
Comment (erimatnor) [reply sql/CMakeLists.txt]: I'll add `${CMAKE_CURRENT_BINARY_DIR}` for consistency and be sure.

File: src/CMakeLists.txt (hunk 3 above)
Comment (erimatnor) [reply src/CMakeLists.txt]: I don't know any reason why it would be a problem. Do you?

File: src/CMakeLists.txt (hunk 4 above)
Comment (erimatnor) [reply src/CMakeLists.txt]: We don't support things that aren't UNIX right now so I don't see a problem. The install target creates a symbolic link to `timescaledb.so`. I can change that to copying `timescaledb-<version>.so` to `/install/path/timescaledb.so`, but for all important purposes it is equivalent to using a symlink. I really think we are splitting hairs here, especially since we want versioned module files and this PR refactors the build system to do a lot of things differently anyway.

**Comment by archenroot** [comment]
@erimatnor superior, this will also simplify build for Gentoo Linux system. Gentoo packaging support so called USE flags, with such USE flag defined you can enable/disable specific functionality of the package to be or not compiled with. Could you provide information about such major features if any?

Just to give you idea, here is for example vanila PostgreSQL package USE flags available:
https://wiki.gentoo.org/wiki/PostgreSQL/QuickStart#USE_flags

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  VERBATIM)
+
+add_library(${PROJECT_NAME} MODULE ${SOURCES} ${HEADERS} ${GITCOMMIT_H})
+
+install(
+  TARGETS ${PROJECT_NAME}
+  DESTINATION ${PG_PKGLIBDIR})
+
+if (UNIX)
+  set_target_properties(${PROJECT_NAME} PROPERTIES
```
Comment (erimatnor) [inline src/CMakeLists.txt]: @cevian FWIW, this only applies to UNIX targets now.

File: src/CMakeLists.txt (hunk 3 above)
Comment (archenroot) [reply src/CMakeLists.txt]: So there is nothing to be defined as USE flag looks like, it makes it even simplier, what I was asking especially is something like OpenCV package which has OCV_OPTIONs to be enabled/disable to support different kind of features (gpu, protocols, codecs support), for reference:
https://github.com/opencv/opencv/blob/master/CMakeLists.txt

Once merged I will produce happily Gentoo ebuild and request to put it into official repos :-)

**Comment by erimatnor** [comment]
@archenroot right now we don't really have any optional features. But, if we ever have optional features in the future, it would certainly be possible to map USE flags in Gentoo builds to the corresponding CMake configuration at build time.

File: src/CMakeLists.txt (hunk 3 above)
Comment (archenroot) [reply src/CMakeLists.txt]: @cevian - compilation with -g could/should be option (ON or OFF - I think better to have OFF by default)

File: src/CMakeLists.txt (hunk 6 above)
Comment (cevian) [reply src/CMakeLists.txt]: I still don't like have two differently named .so installed. Can we not include this change in this PR?
